don't require network access for running tests

When trying to build `grafana-agent` in a sandbox with no network access, I realized the unit tests try to download things from a GCP bucket:

```
=== RUN   TestEnqueueRequestForSelector
    selector_eventhandler_test.go:48:
                Error Trace:    selector_eventhandler_test.go:170
                                                        selector_eventhandler_test.go:48
                Error:          Received unexpected error:
                                Get "https://storage.googleapis.com/kubebuilder-tools/kubebuilder-tools-1.19.2-linux-amd64.tar.gz": dial tcp: lookup storage.googleapis.com on [::1]:53: read udp [::1]:52238->[::1]:53: read: connection refused
                Test:           TestEnqueueRequestForSelector
                Messages:       failed to download dependencies for envtest
--- FAIL: TestEnqueueRequestForSelector (0.00s)
FAIL
FAIL    github.com/grafana/agent/pkg/operator   0.051s
FAIL
```

Can you ship these tools in the repo (maybe via git-lfs), or just require them to be in $PATH?

And make these kind of tests opt-in, or skip if the tools can't be downloaded?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,8 @@
 
 - [FEATURE] Support autologging span attributes as log labels (@mapno)
 
+- [FEATURE] Put Tests requiring Network Access behind a -online flag (@flokli)
+
 - [ENHANCEMENT] The Grafana Agent Operator will now default to deploying
   the matching release version of the Grafana Agent instead of v0.14.0.
   (@rfratto)

--- a/Makefile
+++ b/Makefile
@@ -218,8 +218,8 @@ lint:
 # for packages that have known race detection issues
 # Run tests with -online flag set to true, to include tests requiring network connectivity in CI
 test:
-	CGO_ENABLED=1 go test $(CGO_FLAGS) -race -cover -coverprofile=cover.out -p=4 ./... -online
-	CGO_ENABLED=1 go test $(CGO_FLAGS) -cover -coverprofile=cover-norace.out -p=4 ./pkg/integrations/node_exporter ./pkg/logs -online
+	CGO_ENABLED=1 go test $(CGO_FLAGS) -race -cover -coverprofile=cover.out -p=4 ./... -- -online
+	CGO_ENABLED=1 go test $(CGO_FLAGS) -cover -coverprofile=cover-norace.out -p=4 ./pkg/integrations/node_exporter ./pkg/logs -- -online
 
 clean:
 	rm -rf cmd/agent/agent

--- a/Makefile
+++ b/Makefile
@@ -216,9 +216,10 @@ lint:
 
 # We have to run test twice: once for all packages with -race and then once more without -race
 # for packages that have known race detection issues
+# Run tests with -online flag set to true, to include tests requiring network connectivity in CI
 test:
-	CGO_ENABLED=1 go test $(CGO_FLAGS) -race -cover -coverprofile=cover.out -p=4 ./...
-	CGO_ENABLED=1 go test $(CGO_FLAGS) -cover -coverprofile=cover-norace.out -p=4 ./pkg/integrations/node_exporter ./pkg/logs
+	CGO_ENABLED=1 go test $(CGO_FLAGS) -race -cover -coverprofile=cover.out -p=4 ./... -online
+	CGO_ENABLED=1 go test $(CGO_FLAGS) -cover -coverprofile=cover-norace.out -p=4 ./pkg/integrations/node_exporter ./pkg/logs -online
 
 clean:
 	rm -rf cmd/agent/agent

--- a/pkg/operator/selector_eventhandler_test.go
+++ b/pkg/operator/selector_eventhandler_test.go
@@ -6,6 +6,7 @@ package operator
 
 import (
 	"context"
+	"flag"
 	"fmt"
 	"path/filepath"
 	"runtime"
@@ -34,11 +35,16 @@ var (
 		runtime.GOOS,
 		runtime.GOARCH,
 	)
+
+	online = flag.Bool("online", false, "Run tests requiring network access")
 )
 
 // TestEnqueueRequestForSelector creates an example Kubenretes cluster and runs
 // EnqueueRequestForSelector to validate it works.
 func TestEnqueueRequestForSelector(t *testing.T) {
+	if !*online {
+		t.Skip("skipping test, pass -online to run tests requiring network access")
+	}
 	l := log.NewNopLogger()
 
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)

